### DISCUSSION TITLE
4.0.0rc2

### DISCUSSION THREAD

**Discussion by michael-s-molina** [apache/superset#27844] [General]
Hello Superset Community,

Superset 4.0.0 RC2 is up for testing. Check the [Testing a Release Candidate (RC)](https://github.com/apache/superset/wiki/Release-Process#testing-a-release-candidate-rc) section of our [Release Process](https://github.com/apache/superset/wiki/Release-Process) for information on how to test and report issues with the release candidate.

The release notes:
https://github.com/apache/superset/tree/master/RELEASING/release-notes-4-0

The release candidate:
https://dist.apache.org/repos/dist/dev/superset/4.0.0rc2/

The Git tag for the release:
https://github.com/apache/superset/tree/4.0.0rc2

The change log for the release:
https://github.com/apache/superset/blob/4.0.0rc2/CHANGELOG/4.0.0.md

The instructions for updating to the release:
https://github.com/apache/superset/blob/4.0.0rc2/UPDATING.md

Public keys are available at:
https://www.apache.org/dist/superset/KEYS

Thanks,
The Apache Superset Team

**Comment by sadpandajoe** [comment]
Still seeing this issue in rc2: https://github.com/apache/superset/issues/27500.  Not urgent enough to downvote the release.

**Reply by michael-s-molina** [reply]
Agreed. Not a release blocker.

**Comment by nigzak** [comment]
Hi

you should include in V4 the following security updates:

- https://github.com/apache/superset/pull/27250   update redis dependency
![image](https://github.com/apache/superset/assets/102737855/c248b11d-5811-4d25-a4ae-534dfd9f7c66)
- https://github.com/apache/superset/pull/27405   update setuptools, based on https://github.com/apache/superset/issues/26991
![image](https://github.com/apache/superset/assets/102737855/16d57eeb-e32e-4a75-8cc2-5356c2ef893f)
- certifi package should be updated, 2023.5.7 is outdated and has a high finding, latest is certifi 2024.2.2
![image](https://github.com/apache/superset/assets/102737855/38be8704-9586-4106-87a2-55b7194183dd)



/////
The following findings (high or critical) can NOT be fixed as of now (02.04.2024) in V4 because of ... 
/////

- Debian Expat: no fix yet
![image](https://github.com/apache/superset/assets/102737855/22e67898-9526-48c3-8a25-ed63ae791161)
- PIP: no fix yet
![image](https://github.com/apache/superset/assets/102737855/3c5148c0-fdef-4f20-8db3-55b6420fffc3)
- Selenium: planned with V5, includes breaking changes
![image](https://github.com/apache/superset/assets/102737855/8ce0b54e-7fb6-4bde-a7c2-6aa878268cb7)

**Reply by michael-s-molina** [reply]
Hi @nigzak. We don't include package bumps as part of the updates given the high number of updates. We do publish the [CVEs](https://superset.apache.org/docs/security/cves/) though 😉

**Reply by thinkh** [reply]
@michael-s-molina may I ask, when and how dependencies are updated if not with a  Superset update (i.e., new Superset version)?

**Reply by michael-s-molina** [reply]
@thinkh The dependencies are [updated](https://github.com/apache/superset/compare/3.1.2rc1...4.0.0rc2). We just don't highlight them in the release notes given the number of updates.

**Reply by thinkh** [reply]
Got it. Thanks for the clarification.

**Reply by nigzak** [reply]
Hint: the screenshots and comments are based on the actual dockerhub image 4.0.0RC2 - I made this comment because it is not part of RC2 and I wanted to avoid that already fixed things (or possible easy fixable things) are (again) missed in a first V4 RC3/stable release. 
I would not have put this here if it would be updated already in RC2 - because of this I waited for next RC release and checked this (and not in "older" RC1).
I did not want to highlight some special updates somewhere in Superset V2/V3 with this - I only wanted to ensure it is also fixed in a next V4 version.
If there is a process that it is fixed in a V4.0 stable - all the better :)

**Reply by michael-s-molina** [reply]
@nigzak Now I understood. Sorry for the confusion 😄.

Major releases are time-based as you can see in our [release process](https://github.com/apache/superset/wiki/Release-Process#schedule). That means that some fixes/features might miss the cut and will need to wait for the next release. This is a trade off of having a constant cadence of releases.

**Reply by nigzak** [reply]
Hi @michael-s-molina 

thanks for your explanation.
Sadly this again means that I am not allowed to use a V4.0.0 because the security scans in other systems (JFROG, AWS, ...) are partly marking things even worse to "critical" than in my screenshot - and I am not allowed to deploy things with "high" or "critical" official without long (often senseless) discussions ...
This is also a reason why I cannot migrate as example to Superset 3.1.1 and I must stay on 3.0.x and have to wait for a next update (3.1.x has an additional critical CVE for which I made the pull request and wait for a next stable release).

I know that this might make sense for some reasons in your process to avoid that you on each iteration have to do many additional works and tests (and more).
Still it might (?) would make sense for CVEs (I am not talking about feature changes) to still include this in iterations - if there are no breaking changes (?) - but for sure that might be a big change to the process and I don't know all of your dependencies which might break it so I can only talk from my side about this ...
At least you have currently not a "critical" CVE marked by docker scout as you see in my screenshots - but as written up that is only one database source for CVEs and each system marks it slightly different (higher/lower)

At least for me that confirms that I sadly never ever can use a x.0.0 version ...
I hope at least that the done fixes then are re-integrated automatically in the x.0.1 versions (?)

**Reply by michael-s-molina** [reply]
Hi @nigzak. These are valid concerns. @dpgaspar is the one leading the efforts regarding CVEs so I suggest pairing up with him to see if we can mitigate some of the issues you are raising. In terms of releases, every security fix that is merged to master is always included in a patch/minor/major release depending on the changes. In other words, as long as you resolve the issues on master, they will eventually end up in a release. What we cannot do is block the releases until we resolve all possible third-party CVEs.

**Comment by sadpandajoe** [comment]
Just tried to do a migration and ran into this error. Not a blocker for us since most of our workspaces have been updated but others may run into this problem.

https://github.com/apache/superset/issues/27885

**Reply by michael-s-molina** [reply]
Issue closed. Check this [comment](https://github.com/apache/superset/issues/27885#issuecomment-2037144428) for more context.
